crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs

```diff
@@ -0,0 +1,137 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall};
+use ruff_python_semantic::Modules;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+#[derive(Debug, Eq, PartialEq)]
+enum Replacement {
+    None,
+    Name(String),
+}
+
+/// ## What it does
+/// Checks for uses of deprecated Airflow functions and values.
+///
+/// ## Why is this bad?
+/// Airflow 3.0 removed various deprecated functions, members, and other
+/// values. Some have more modern replacements. Others are considered too niche
+/// and not worth to be maintained in Airflow.
+///
+/// ## Example
+/// ```python
+/// from airflow.utils.dates import days_ago
+///
+///
+/// yesterday = days_ago(today, 1)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from datetime import timedelta
+///
+///
+/// yesterday = today - timedelta(days=1)
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct Airflow3Removal {
+    deprecated: String,
+    replacement: Replacement,
+}
+
+impl Violation for Airflow3Removal {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Airflow3Removal {
+            deprecated,
+            replacement,
+        } = self;
+        match replacement {
+            Replacement::None => format!("`{deprecated}` is removed in Airflow 3.0"),
+            Replacement::Name(name) => {
+                format!("`{deprecated}` is removed in Airflow 3.0; use {name} instead")
+            }
+        }
+    }
+}
+
+struct DeprecatedArgument(Vec<&'static str>, Option<&'static str>);
+
+fn removed_argument(checker: &mut Checker, qualname: &QualifiedName, arguments: &Arguments) {
+    let deprecations = match qualname.segments() {
+        ["airflow", .., "DAG" | "dag"] => vec![DeprecatedArgument(
+            vec!["timetable", "schedule_interval"],
+            Some("schedule"),
+        )],
```

Comment:
Returning a `vec` of possible deprecations here is suboptimal for performance because it means that Ruff allocates at least two vectors for every `airflow` call expression. 

I think we could do better by changing the API slightly

```rust
fn deprecated_keyword_argument(arguments: &Arguments, argument: &str, replacement: Option<&str>) -> Option<Diagnostic> {
	let keyword = arguments.find_keyword(argument)?;
	
	Some(Diagnostic::new(
        Airflow3Removal {
            deprecated: argument.to_string(),
            replacement: match replacement {
                Some(name) => Replacement::Name(name.to_owned()),
                None => Replacement::None,
            },
        },
        keyword
            .arg
            .as_ref()
            .map_or_else(|| keyword.range(), Ranged::range),
    ));
}
```

and you can then use it like this

```rust
    let deprecations = match qualname.segments() {
        ["airflow", .., "DAG" | "dag"] => {
					checker.diagnostics.extend(removed_argument(arguments, "timetable", Some("schedule")));
					checker.diagnostics.extend(removed_argument(arguments, "schedule_interval", Some("schedule")));
		}
```

Reply:
Sounds good to me. I was struggling to find a way to do this more efficiently.

Reply:
Alright this is pretty nice! Need to suppress Clippy since we do want to add more cases here in the future.